### DISCUSSION TITLE
Avalonia Cross Platform use of Pseudo Classes for Android and IOS Development.

### DISCUSSION THREAD

**Discussion by KentSwan** [AvaloniaUI/Avalonia#14306] [General]
I've run into a conundrum.  First when developing for Android and using the local simulator (1) it appears that pointerover pseudo class has no effect. (2) on the simulator on a conventional monitor, how do you interact with it as if you were using your finger?  Select seems to work for input controls but how is drag and drop, swipe, etc manged?

**Comment by thevortexcloud** [comment]
I don't think the android simulator does mouse hover simulation by default? Not sure about iOS. It however has been a number of years since I took a proper look at mobile development.

> But how is drag and drop, swipe, etc manged?


Avalonia has gesture recongnisers. EG https://docs.avaloniaui.net/docs/concepts/input/gestures

It depends on exactly what you are looking for. There are some third party controls that do try to make mobile work more "seamlessly". As not all of the official controls work very well on a mobile device.

**Reply by maxkatz6** [reply]
Yeah, android simulator tried to "simulate" touch input. Same for iOS. 
So, it's not possible to get "pointerover" there, as touch input doesn't have "pointer" to hover with.

**Reply by thevortexcloud** [reply]
Out of curiosity: android has full mouse support. You can plug a mouse into most phones via the USB port and it will work. I have known people who have done it with tablets and used it as their desktop. Does Avalonia support it in that scenario?

In any case, I imagine some sort of long press selector/event might be useful here.

**Reply by maxkatz6** [reply]
Yes, Avalonia supports full mouse input on Android (and on iOS from recent nightly builds). But not with simulator.

**Reply by thevortexcloud** [reply]
Very nice.

**Reply by KentSwan** [reply]
Thanks all.  So basically in the emulator so far all I've been able to verify is mouse select for an input field. I need to research and experiment to see what other minimal mouse features like mouse pointer, move, and of course, drag and drop work.   I will want to use a live Android device with adb for live testing.  

For my application, I can use a phone but prefer a modest-sized tablet (for on screen keyboard). It's going to need a decent resolution autofocus camera more or less like our modern cell phones along with both  Bluetooth and WiFi and, of course, a reliable adb developer debug capability. I'm open to suggestions.

**Reply by thevortexcloud** [reply]
It's been a _long_ time since I have looked into android tablets but from memory Samsung is the only company to make actual decent android tablets. Everything else is kind of cheap and terrible. But the market has possibly changed since I last looked. Google for example apparently has their own tablet now. Pixel phones for example are fairly open, to the point you can just replace the entire OS. But not sure if you can do that with their tablet.

**Comment by KentSwan** [comment]
Thanks for the pointers

**Reply by thevortexcloud** [reply]
I should mention if you want your device to actually stay up to date with new android versions (which is fairly useful for development purposes), get a google device. They are now promising 7 years of major updates for the Pixel 8. Which is almost unheard of in the android world. Most devices get 3 at best.

https://support.google.com/pixelphone/answer/4457705
